markets/perps-market/test/integration/Orders/OffchainAsyncOrder.fees.test.ts

```diff
@@ -1,6 +1,420 @@
+              ...commonOpenPositionProps,
+              sizeDelta,
+              price: ethPrice,
+            });
+          });
+
+          it('validate fees paid on short position', async () => {
+            const { traderBalance, perpMarketWithdrawable, keeperBalance } = await getBalances();
+
+            assertBn.equal(
+              traderBalance,
+              balancesAfterLong.traderBalance
+                .sub(feesPaidOnShort.totalFees)
+                .add(balancesAfterLong.accountPnl)
+            );
+            assertBn.equal(
+              perpMarketWithdrawable,
+              balancesAfterLong.perpMarketWithdrawable
+                .add(feesPaidOnShort.perpsMarketFee)
+                .sub(balancesAfterLong.accountPnl)
+            );
+            assertBn.equal(
+              keeperBalance,
+              balancesAfterLong.keeperBalance.add(feesPaidOnShort.keeperFee)
+            );
+          });
+        });
+
+        describe('flip the order side (maker + taker)', () => {
+          let feesPaidOnShort: ReturnType<typeof computeFees>;
+          const sizeDelta = bn(-3); // original size is 2, reduce by 3 => flip to short -1
+
+          before(restoreToLongOrder);
+
+          before('open a large short order', async () => {
+            feesPaidOnShort = computeFees(
+              initialLongSize,
+              sizeDelta,
+              await systems().PerpsMarket.fillPrice(ethMarketId, sizeDelta, ethPrice)
+            );
+
+            await openPosition({
+              ...commonOpenPositionProps,
+              sizeDelta,
+              price: ethPrice,
+            });
+          });
+
+          it('validate fees paid', async () => {
+            const { traderBalance, perpMarketWithdrawable, keeperBalance } = await getBalances();
+
+            assertBn.equal(
+              traderBalance,
+              balancesAfterLong.traderBalance.sub(feesPaidOnShort.totalFees)
+            );
+            assertBn.equal(
+              perpMarketWithdrawable,
+              balancesAfterLong.perpMarketWithdrawable.add(feesPaidOnShort.perpsMarketFee)
+            );
+            assertBn.equal(
+              keeperBalance,
+              balancesAfterLong.keeperBalance.add(feesPaidOnShort.keeperFee)
+            );
+          });
+        });
+      });
+    });
+  }
+
+  const getBalances = async () => {
+    const traderBalance = await systems().PerpsMarket.totalCollateralValue(2);
+    const perpMarketWithdrawable = await systems().Core.getWithdrawableMarketUsd(ethMarketId);
+    const keeperBalance = await systems().USD.balanceOf(keeper().getAddress());
+    const accountPnl = (await systems().PerpsMarket.getOpenPosition(2, ethMarketId))[0];
+    return {
+      traderBalance,
+      perpMarketWithdrawable,
+      keeperBalance,
+      accountPnl,
+    };
+  };
+
+  const computeFees: (
+    sizeBefore: ethers.BigNumber,
+    sizeDelta: ethers.BigNumber,
+    price: ethers.BigNumber
+  ) => {
```

Comment:
generally, ive been trying not to write functions that mimic contract that calculate the value we're wanting to assert.  if there's a bug in the contract and the function mimics what hte contract is doing, then teh bug will be hidden.  i've resorted to having excel sheet with what i think the calc is, and then having afif and all of you checking the logic to make sure it all looks right.  and if those values match up with what the contract is returning then i feel confident.

so just to wary of that; not saying this function isn't returning the right value but yeah hope that makes sense 👍🏽

Reply:
tbh, I prefer our code to have the values of the tests calculated in place, otherwise a minimun change in the behavior will need to change all hardcoded values. Also, if somebody doesn't have access to the sheet, it cannot verify the values makes sense. 
In this case I went the other way round. I did the sheet, the test helper and checked the values made sense. In fact with that function is where I found we had a rounding issue on the calculations and made the function kind of mimic the code of the test.

Reply:
yea fair points.  the sheets is an experiement for sure and not scalable.  you might have convinced me to have _some_ form of code to determine values.   i'll keep iterating; thanks